src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -585,18 +594,17 @@ InProgressState inProgressState
                     }
 
                     async Task<FinalState> BuildFinalStateFromInProgressStateAsync(
-                        SolutionCompilationState compilationState, InProgressState state, CancellationToken cancellationToken)
+                        SolutionCompilationState compilationState, InProgressState inProgressState, CancellationToken cancellationToken)
                     {
                         try
                         {
-                            var (compilationWithoutGenerators, compilationWithGenerators, generatorDriver) = await BuildDeclarationCompilationFromInProgressAsync(
-                                state, cancellationToken).ConfigureAwait(false);
+                            var allSyntaxTreesParsedState = await BuildAllSyntaxTreesParsedStateFromInProgressStateAsync(
+                                inProgressState, cancellationToken).ConfigureAwait(false);
 
                             return await FinalizeCompilationAsync(
                                 compilationState,
-                                compilationWithoutGenerators,
-                                state.GeneratorInfo.WithDriver(generatorDriver),
-                                compilationWithGenerators,
+                                allSyntaxTreesParsedState,
+                                allSyntaxTreesParsedState.CompilationWithGeneratedDocuments,
```

Comment:
Good insight that FinalizeCompilation can always take an AllSyntaxTreesParsed state.

Reply:
yes.  this really helped clear things up for me.